pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -161,15 +159,21 @@ func (c *defaultMaintenanceControl) Maintain(shootObj *gardenv1beta1.Shoot, key
 		return nil
 	}
 
-	// Check if the CloudProfile contains another version of the machine image.
-	machineImageFound, machineImage, err := helper.DetermineMachineImage(*operation.Shoot.CloudProfile, operation.Shoot.GetMachineImage().Name)
+	// make sure to base decision upon up to date cloud profile (lister in operation might not have synced yet)
+	cloudProfileForShoot, err := c.k8sGardenClient.Garden().GardenV1beta1().CloudProfiles().Get(shoot.Spec.Cloud.Profile, metav1.GetOptions{});
```

Comment:
Yes happened to me in the integration tests. 
This could happen any time when we update the cloud profile when a maintenance operation on a shoot is about to start - and then the maintenance decision is being made upon an invalid data basis.

I guess it is an easy fix to be sure. 
I we decide the performance penalty is too high, its also ok, but it might happen that a shoot wrongly gets down/upgraded until the next maintenance cycle.

Reply:
I get your point but there is some point in time you have to rely on the state that is inside the `CloudProfile`.

Even if you refetch it via the client, it can still be changed right after that.

Reply:
I agree with @timuthy. Also, assume the following sequence:

```
time 1, the expiration time is set to value A
time 2-10, some shoots are maintained with the given value A
time 11, the operator extends/changes the expiration time to value B AND some arbitrary shoot X is maintained at the same time (with value A)
```

It doesn't matter whether the shoot X is still maintained with value A because other shoots have already been maintained with the same A previously as well. We cannot ensure consistency anyways, and we don't get any benefit of reading having one shoot less not maintained with the updated value.